include/crm/common/internal.h

```diff
@@ -265,6 +265,48 @@ extern int pcmk__score_red;
 extern int pcmk__score_green;
 extern int pcmk__score_yellow;
 
+/*!
+ * \internal
+ * \brief Allocate new zero-initialized memory, asserting on failure
+ *
+ * \param[in] file      File where \p function is located
+ * \param[in] function  Calling function
+ * \param[in] line      Line within \p file
+ * \param[in] nmemb     Number of elements to allocate memory for
+ * \param[in] size      Size of each element
+ *
+ * \return Newly allocated memory of of size <tt>nmemb * size</tt> (guaranteed
+ *         not to be \c NULL)
+ *
+ * \note The caller is responsible for freeing the return value using \c free().
+ */
+static inline void *
+pcmk__assert_alloc_as(const char *file, const char *function, uint32_t line,
+                      size_t nmemb, size_t size)
+{
+    void *ptr = calloc(nmemb, size);
+
+    if (ptr == NULL) {
+        crm_abort(file, function, line, "Out of memory", FALSE, FALSE);
```

Comment:
> FYI, despite the argument, this dumps core via `abort()`. I wonder if we should do `crm_abort(..., FALSE, TRUE)` plus `crm_exit()`.

Ah, you're right. `CRM_LOG_ASSERT()` basically fits the bill except that it stringifies the expression instead of using "Out of memory".

However, that does raise a question: `CRM_LOG_ASSERT()` messes with a libqb callsite:
```
#  define CRM_LOG_ASSERT(expr) do {                 \
        if (!(expr)) {                                                  \
            static struct qb_log_callsite *core_cs = NULL;              \
            if(core_cs == NULL) {                                       \
                core_cs = qb_log_callsite_get(__func__, __FILE__,       \
                                              "log-assert", LOG_TRACE,  \
                                              __LINE__, 0);             \
            }                                                           \
            crm_abort(__FILE__, __func__, __LINE__, #expr,              \
                      core_cs?core_cs->targets:FALSE, TRUE);            \
        }                                                               \
    } while(0)
```

Do you know off hand what this achieves? I think I figured it out after I typed this question; see next paragraphs and please confirm or clarify if possible. And do we need something like it in `pcmk__mem_assert()`? I think not, because if I understand it correctly, the intent is to dump core for detailed info if tracing; no reason to ever dump core for out-of-memory.

I see that it creates a callsite at trace level. If it gets one, and any targets are enabled for the callsite (however that even happens), it sets `do_core=TRUE`. That causes a core to be dumped -- which is probably the intent, for tracing purposes. However, if the process isn't a daemon, then `do_fork` is ignored: the program itself aborts instead of forking a child to abort. I would never expect `CRM_LOG_ASSERT()` to cause a program to abort.

Otherwise, if there's no callsite or no targets are enabled, `CRM_LOG_ASSERT()` passes `FALSE` (which, combined with `do_fork=TRUE`, says "log without dumping core"). Which is what I would expect `CRM_LOG_ASSERT()` to do in all cases.

`CRM_CHECK()` is set up the same way.

> Also FYI, when do_fork is TRUE, `crm_abort()` doesn't always abort

*Something* aborts. The process itself if it's not a daemon, or a forked child if it is. Good point to be aware of though.

> At one point I had planned to replace `crm_abort()` with something saner but never got around to it

I think that's worthwhile, but hopefully it can be postponed. I'm already way farther down a rabbit hole than I want to be.